docs/rfcs/025-generation-numbers.md

```diff
@@ -0,0 +1,650 @@
+
+Item 2. is a non-issue operationally: it's harmless to delay deletions, the only impact of objects pending deletion is
+the S3 capacity cost.
+
+Item 3. is an issue if safekeepers are low on disk space and the control plane is unavailable for a long time.
+
+For a managed service, the general approach should be to make sure we are monitoring & respond fast enough
+that control plane outages are bounded in time. The separate of console and control plane will also help
+to keep the control plane itself simple and robust.
+
+We should also implement an "escape hatch" config for node generation numbers, where in a major disaster outage,
+we may manually run pageservers with a hand-selected generation number, so that we can bring them online
+independently of a control plane.
+
+### Rollout
+
+Although there is coupling between components, we may deploy most of the new data plane components
+independently of the control plane: initially they can just use a static generation number.
+
+#### Phase 1
+
+The pageserver is deployed with some special config to:
+
+- Always act like everything is generation 1 and do not wait for a control plane issued generation on startup.
+- Skip the places in deletion and remote_consistent_lsn updates where we would call into control plane
+
+The storage broker will tolerate the timeline state omitting generation numbers.
+
+The safekeeper will be aware of both new and old versions of `PageserverFeedback` message, and tolerate
+the old version.
+
+#### Phase 2
+
+The control plane changes are deployed: control plane will now track and increment generation numbers.
+
+#### Phase 3
+
+The pageserver is deployed with its control-plane-dependent changes enabled: it will now require
+the control plane to issue a generation number, and require to communicate with the control plane
+prior to processing deletions.
+
+### On-disk backward compatibility
+
+Backward compatibility with existing data is straightforward:
+
+- When reading the index, we may assume that any layer whose metadata doesn't include
+  generations will have a generation-less key path.
+- When locating the index file on attachment, we may use the "fallback" listing path
+  and if there is only a generation-less index, that is the one we load.
+
+It is not necessary to re-write an existing layers: even new index files will be able
+to represent generation-less layers.
+
+### On-disk forward compatibility
+
+We will do a two phase rollout, probably over multiple releases because we will naturally
+have some of the read-side code ready before the overall functionality is ready:
+
+1. Deploy pageservers which understand the new index format and generation numbers
+   in keys, but do not write objects with generation numbers in the keys.
+2. Deploy pageservers that write objects with generation numbers in the keys.
+
+Old pageservers will be oblivious to generation numbers. That means that they can't
+read objects with generation numbers in the name.
```

Comment:
That sentence is the problem that is being solved by the two-phased rollout above, right?

Reply:
Right, clarified 6d59f5c1